mlx/ops.cpp

```diff
@@ -2698,6 +2698,39 @@ array gather(
       inputs);
 }
 
+array kron(array& a, array& b, StreamOrDevice s /* = {} */) {
+  auto out_type = promote_types(a.dtype(), b.dtype());
+
+  if (a.size() == 0 || b.size() == 0) {
+    throw std::invalid_argument(
+        "Input arrays cannot be empty."); // Or your mx equivalent
+  }
+  while (a.ndim() < b.ndim()) {
+    a = expand_dims(a, {0}, s);
+  }
+  while (b.ndim() < a.ndim()) {
+    b = expand_dims(b, {0}, s);
+  }
+  int ndim = a.ndim();
+  std::vector<int> a_new_shape;
+  std::vector<int> b_new_shape;
+  std::vector<int> out_shape;
+
+  for (int i = 0; i < ndim; ++i) {
+    a_new_shape.push_back(a.shape()[i]);
+    a_new_shape.push_back(1);
+    b_new_shape.push_back(1);
+    b_new_shape.push_back(b.shape()[i]);
+    out_shape.push_back(a.shape()[i] * b.shape()[i]);
+  }
+
+  array a_reshaped = reshape(a, a_new_shape, s);
+  array b_reshaped = reshape(b, b_new_shape, s);
+
+  array result = multiply(a_reshaped, b_reshaped, s);
+  return reshape(result, out_shape, s);
```

Comment:
This is completely fine and it might be a personal preference and nitpick but just in case there is a future optimization that can auto-move the intermediates I would simply write it as follows

```cpp
return reshape(
    multiply(
        reshape(a, std::move(a_new_shape), s),
        reshape(b, std::move(b_new_shape), s),
        s),
    std::move(out_shape),
    s);
```